src/daemon/ca/status.rs

```diff
@@ -149,10 +180,36 @@ impl StatusStore {
         published: Vec<PublishElement>,
         next_hours: i64,
     ) -> KrillResult<()> {
+        self.update_ca_status(ca, |status| status.repo.set_published(uri, published, next_hours))
+            .await
+    }
+
+    async fn update_ca_status<F>(&self, ca: &Handle, op: F) -> KrillResult<()>
+    where
+        F: FnOnce(&mut CaStatus),
+    {
         let _lock = self.lock.write().await;
         let mut status = self.get_ca_status(ca)?;
-        status.repo.set_published(uri, published, next_hours);
-        self.set_ca_status(ca, &status)
+        op(&mut status);
+
+        self.store.store(&Self::status_key(ca), &status)?;
+
+        Ok(())
+    }
+
+    async fn update_ca_child_status<F>(&self, ca: &Handle, child: &ChildHandle, op: F) -> KrillResult<()>
+    where
+        F: FnOnce(&mut ChildStatus),
+    {
+        self.update_ca_status(ca, |status| {
+            if !status.children.contains_key(child) {
```

Comment:
Could this potentially suffer from a race condition whereby two thread both try to check for the presence of the key at the same time and then both try to insert it, or is that prevented by the acquisition of the `self.lock` in `update_ca_status()` ?

Perhaps it would make sense to use [`Entry::or_insert_with()`](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with) anyway?

Reply:
This is safely inside a closure in a lock. The downside of `entry()` is that we always need to clone the key. But, we were getting the value twice. I have pushed code which will only insert and clone if they value could not be retrieved in the first place.